File: frontend/src/components/benchmark/content/floyd_warshall/implementations/floyd_warshall.go
package main

import (
    "math"
)

type FloydWarshall struct {
    nodes    map[int]map[int]bool
    dist     map[int]map[int]int
    nextNode map[int]map[int]int
}

func NewFloydWarshall() *FloydWarshall {
    return &FloydWarshall{
        nodes:    make(map[int]map[int]bool),
        dist:     make(map[int]map[int]int),
        nextNode: make(map[int]map[int]int),
    }
}

func (fw *FloydWarshall) AddEdge(u, v, weight int) {
    if fw.nodes[u] == nil {
        fw.nodes[u] = make(map[int]bool)
    }
    if fw.nodes[v] == nil {
        fw.nodes[v] = make(map[int]bool)
    }
    fw.nodes[u][v] = true
    fw.nodes[v][u] = true

    if fw.dist[u] == nil {
        fw.dist[u] = make(map[int]int)
    }
    if fw.dist[v] == nil {
        fw.dist[v] = make(map[int]int)
    }
    fw.dist[u][v] = weight
    fw.dist[v][u] = weight

    if fw.nextNode[u] == nil {
        fw.nextNode[u] = make(map[int]int)
    }
    if fw.nextNode[v] == nil {
        fw.nextNode[v] = make(map[int]int)
    }
    fw.nextNode[u][v] = v
    fw.nextNode[v][u] = u
}

func (fw *FloydWarshall) AddDirectedEdge(u, v, weight int) {
    if fw.nodes[u] == nil {
        fw.nodes[u] = make(map[int]bool)
    }
    if fw.nodes[v] == nil {
        fw.nodes[v] = make(map[int]bool)
    }
    fw.nodes[u][v] = true

    if fw.dist[u] == nil {
        fw.dist[u] = make(map[int]int)
    }
    if fw.dist[v] == nil {
        fw.dist[v] = make(map[int]int)
    }
    fw.dist[u][v] = weight

    if fw.nextNode[u] == nil {
        fw.nextNode[u] = make(map[int]int)
    }
    if fw.nextNode[v] == nil {
        fw.nextNode[v] = make(map[int]int)
    }
    fw.nextNode[u][v] = v
}

func (fw *FloydWarshall) BuildFromEdges(edges [][3]int, directed bool) {
    for _, edge := range edges {
        if directed {
            fw.AddDirectedEdge(edge[0], edge[1], edge[2])
        } else {
            fw.AddEdge(edge[0], edge[1], edge[2])
        }
    }
}

func (fw *FloydWarshall) getAllNodes() []int {
    result := []int{}
    for node := range fw.nodes {
        result = append(result, node)
    }
    return result
}

func (fw *FloydWarshall) initialize() {
    allNodes := fw.getAllNodes()

    for _, u := range allNodes {
        if fw.dist[u] == nil {
            fw.dist[u] = make(map[int]int)
        }
        if fw.nextNode[u] == nil {
            fw.nextNode[u] = make(map[int]int)
        }

        for _, v := range allNodes {
            if u == v {
                fw.dist[u][v] = 0
            } else if _, exists := fw.dist[u][v]; !exists {
                fw.dist[u][v] = math.MaxInt32
            }
        }
    }
}

func (fw *FloydWarshall) AllPairsShortestPaths() {
    fw.initialize()
    allNodes := fw.getAllNodes()

    for _, k := range allNodes {
        for _, i := range allNodes {
            for _, j := range allNodes {
                if fw.dist[i][k] != math.MaxInt32 && fw.dist[k][j] != math.MaxInt32 &&
                    fw.dist[i][k]+fw.dist[k][j] < fw.dist[i][j] {
                    fw.dist[i][j] = fw.dist[i][k] + fw.dist[k][j]
                    fw.nextNode[i][j] = fw.nextNode[i][k]
                }
            }
        }
    }
}

func (fw *FloydWarshall) ShortestPath(u, v int) int {
    if fw.dist[u] == nil || fw.dist[u][v] == 0 {
        return math.MaxInt32
    }
    return fw.dist[u][v]
}

func (fw *FloydWarshall) GetPath(u, v int) []int {
    if fw.nextNode[u] == nil || fw.nextNode[u][v] == 0 {
        return []int{}
    }

    path := []int{u}

    for u != v {
        u = fw.nextNode[u][v]
        path = append(path, u)
    }

    return path
}

func (fw *FloydWarshall) GetAllDistances() map[int]map[int]int {
    return fw.dist
}

func (fw *FloydWarshall) HasNegativeCycle() bool {
    allNodes := fw.getAllNodes()

    for _, i := range allNodes {
        if fw.dist[i][i] < 0 {
            return true
        }
    }
    return false
}

File: frontend/src/components/benchmark/content/floyd_warshall/implementations/floyd_warshall.php
<?php

class FloydWarshall {
    private array $nodes = [];
    private array $dist = [];
    private array $nextNode = [];

    public function addEdge(int $u, int $v, int $weight): void {
        if (!isset($this->nodes[$u])) $this->nodes[$u] = [];
        if (!isset($this->nodes[$v])) $this->nodes[$v] = [];
        $this->nodes[$u][$v] = true;
        $this->nodes[$v][$u] = true;

        if (!isset($this->dist[$u])) $this->dist[$u] = [];
        if (!isset($this->dist[$v])) $this->dist[$v] = [];
        $this->dist[$u][$v] = $weight;
        $this->dist[$v][$u] = $weight;

        if (!isset($this->nextNode[$u])) $this->nextNode[$u] = [];
        if (!isset($this->nextNode[$v])) $this->nextNode[$v] = [];
        $this->nextNode[$u][$v] = $v;
        $this->nextNode[$v][$u] = $u;
    }

    public function addDirectedEdge(int $u, int $v, int $weight): void {
        if (!isset($this->nodes[$u])) $this->nodes[$u] = [];
        if (!isset($this->nodes[$v])) $this->nodes[$v] = [];
        $this->nodes[$u][$v] = true;

        if (!isset($this->dist[$u])) $this->dist[$u] = [];
        if (!isset($this->dist[$v])) $this->dist[$v] = [];
        $this->dist[$u][$v] = $weight;

        if (!isset($this->nextNode[$u])) $this->nextNode[$u] = [];
        if (!isset($this->nextNode[$v])) $this->nextNode[$v] = [];
        $this->nextNode[$u][$v] = $v;
    }

    public function buildFromEdges(array $edges, bool $directed = false): void {
        foreach ($edges as $edge) {
            if ($directed) {
                $this->addDirectedEdge($edge[0], $edge[1], $edge[2]);
            } else {
                $this->addEdge($edge[0], $edge[1], $edge[2]);
            }
        }
    }

    private function getAllNodes(): array {
        return array_keys($this->nodes);
    }

    private function initialize(): void {
        $allNodes = $this->getAllNodes();

        foreach ($allNodes as $u) {
            if (!isset($this->dist[$u])) $this->dist[$u] = [];
            if (!isset($this->nextNode[$u])) $this->nextNode[$u] = [];

            foreach ($allNodes as $v) {
                if ($u == $v) {
                    $this->dist[$u][$v] = 0;
                } else if (!isset($this->dist[$u][$v])) {
                    $this->dist[$u][$v] = INF;
                }
            }
        }
    }

    public function allPairsShortestPaths(): void {
        $this->initialize();
        $allNodes = $this->getAllNodes();

        foreach ($allNodes as $k) {
            foreach ($allNodes as $i) {
                foreach ($allNodes as $j) {
                    if ($this->dist[$i][$k] != INF && 
                        $this->dist[$k][$j] != INF &&
                        $this->dist[$i][$k] + $this->dist[$k][$j] < $this->dist[$i][$j]) {
                        $this->dist[$i][$j] = $this->dist[$i][$k] + $this->dist[$k][$j];
                        $this->nextNode[$i][$j] = $this->nextNode[$i][$k];
                    }
                }
            }
        }
    }

    public function shortestPath(int $u, int $v): int {
        if (!isset($this->dist[$u]) || !isset($this->dist[$u][$v])) {
            return INF;
        }
        return $this->dist[$u][$v];
    }

    public function getPath(int $u, int $v): array {
        if (!isset($this->nextNode[$u]) || !isset($this->nextNode[$u][$v])) {
            return [];
        }

        $path = [$u];

        while ($u != $v) {
            $u = $this->nextNode[$u][$v];
            $path[] = $u;
        }

        return $path;
    }

    public function getAllDistances(): array {
        return $this->dist;
    }

    public function hasNegativeCycle(): bool {
        $allNodes = $this->getAllNodes();

        foreach ($allNodes as $i) {
            if ($this->dist[$i][$i] < 0) {
                return true;
            }
        }
        return false;
    }
}

?>